File: LeetCode/java/2501_Longest_Square_Streak_in_an_Array.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int longestSquareStreak(int[] nums) {
        int answer = -1;

        Arrays.sort(nums);

        Set<Integer> set = new HashSet<>();

        for (int n : nums) {
            set.add(n);
        }

        for (int i = 0; i < nums.length; i++) {
            int cnt = 1;
            int n = nums[i];

            while (true) {
                if ((long)n * n > 100000) {
                    break;
                }

                n = n * n;

                if (!set.contains(n)) {
                    break;
                }

                if (set.contains(n)) {
                    cnt++;

                    if (cnt > 1) {
                        answer = Math.max(answer, cnt);
                    }
                }
            }
        }

        return answer;
    }
}

File: LeetCode/cpp/2501_Longest_Square_Streak_in_an_Array.cpp
class Solution {
public:
	int longestSquareStreak(vector<int>& nums) {
		int answer = -1;

		sort(nums.begin(), nums.end());

		set<int> numsSet;

		for (int n : nums) {
			numsSet.insert(n);
		}

		for (int i = 0; i < nums.size(); i++) {
			int cnt = 1;
			int n = nums[i];

			while (1) {
				if ((long long)n * n > 100000) {
					break;
				}

				n = n * n;

				if (numsSet.find(n) == numsSet.end()) {
					break;
				}

				if (numsSet.find(n) != numsSet.end()) {
					cnt++;

					if (cnt > 1) {
						answer = max(answer, cnt);
					}
				}
			}
		}

		return answer;
	}
};